deps/spidershim/src/v8functiontemplate.cc

```diff
@@ -0,0 +1,315 @@
+    }
+
+    protoObj = GetObject(protoInstance.ToLocalChecked());
+    assert(protoObj);
+
+    // Now we have to give it the right prototype, though!
+    if (!JS_SplicePrototype(cx, protoObj, protoProto)) {
+      // XXXbz what to do with the exception?
+      return MaybeLocal<Function>();
+    }
+  }
+
+  // TODO: We should be doing something faster here in terms of setting up
+  // prototypes right from the beginning or something.  Maybe we can refactor
+  // Function::New's implementation a bit to allow us to pass the right things
+  // into there.
+  Local<Function> func = Function::New(context->GetIsolate(),
+                                       callback,
+                                       dataV8Val,
+                                       lengthVal.toInt32());
+  if (!*func) {
+    return MaybeLocal<Function>();
+  }
+
+  // Copy the bits set on us via Template::Set.
+  JS::RootedObject funcObj(cx, GetObject(func));
+  assert(funcObj);
+  if (!JS_CopyPropertiesFrom(cx, funcObj, obj)) {
+    // XXXbz What do we do with the exception?
+    return MaybeLocal<Function>();
+  }
+
+  // TODO: Is the result of all this supposed to have a non-configurable
+  // .prototype like builtins, or a configurable one like scripted functions?
+  // Let's assume the former.
+  if (!JS_LinkConstructorAndPrototype(cx, funcObj, protoObj)) {
+    // XXXbz what to do with the exception?
+    return MaybeLocal<Function>();
+  }
+
+  // And now we're all instantiated.  Cache that value.
+  js::SetReservedSlot(obj, InstanceSlot, *GetValue(func));
+
+  return func;
+}
+
+// TODO: Need to implement InstanceTemplate().  In particular, need to figure
+// out how this interacts with the FunctionCallback we were handed to start
+// with!
+
+Local<ObjectTemplate>
+FunctionTemplate::PrototypeTemplate()
+{
+  // XXXbz the V8 behavior here is weird: calling PrototypeTemplate() will
+  // modify this function template, but they don't ensure that we have not been
+  // instantiated yet...  Ah, well.  Let's just do what they do.
+  Isolate* isolate = Isolate::GetCurrent();
+  JSContext* cx = JSContextFromIsolate(isolate);
+  JS::RootedObject obj(cx, GetObject(this));
+  assert(obj);
+
+  JS::RootedValue protoTemplateVal(cx, js::GetReservedSlot(obj, ProtoSlot));
+  if (protoTemplateVal.isObject()) {
+    // XXX bz is this the right way to do this?
+    Local<Template> templ =
+      internal::Local<Template>::NewTemplate(isolate, protoTemplateVal);
+    return Local<ObjectTemplate>(static_cast<ObjectTemplate*>(*templ));
+  }
+
+  assert(protoTemplateVal.isUndefined());
+  Local<ObjectTemplate> templ = ObjectTemplate::New(isolate);
+  if (!*templ) {
+    return Local<ObjectTemplate>();
+  }
+
+  js::SetReservedSlot(obj, ProtoSlot, JS::ObjectValue(*GetObject(*templ)));
+  return templ;
+}
+
+// TODO: Need to implement SetClassName.  Does this map to JSClass names, and if
+// so do we need to maintain our own (heap-allocated, presumably) instance
+// JSClass?
```

Comment:
The answer to the first question at least seems to be yes.  See `TestFunctionTemplateAccessor` for example where it sets the `"funky"` class name...